* DIVTRANS							   :Muchardt:
  
  We look at the implication of HERV sequences in the reactivation of
  immune related genes.

** DONE Project description					     :ATTACH:
   CLOSED: [2018-01-08 Mon 16:06] DEADLINE: <2018-01-05 Fri>
   :PROPERTIES:
   :Attachments: Enhancers_of_Life.pdf
   :ID:       0f6a94ab-cd00-44b4-8a39-3ed6bf8f3414
   :END:      
Mapping of Enhancers from transcriptome data Enhancers of
transcription are regulatory sequences enabling gene expression from a
distance. The landscape of active enhancers is cell-type specific and
provides extensive information on the transcription factors at
play. Currently, enhancers are mostly mapped based on the histone
modifications positioned on their DNA sequence. This type of data is
abundantly available for tumor-derived tissue culture cells, but
difficult to obtain when the biological material has a limited
availability. As an alternative approach, it is possible to detect
enhancers as sites of divergent transcription. The objective of this
project will be to develop tools allowing detection of sites of
divergent transcription in transcriptome and run-on data, to evaluate
the quality of the prediction by comparing the outcome with existing
enhancer maps, and then ultimately use this approach to identify
changes in the enhancer landscape between patients with multiple
sclerosis and healthy controls.

The endpoint of the project is to enable us to use RNA-seq data to
identify transcription factor activities and pathways modified in
patients with multiple sclerosis. The strategy relies on the recent
observation that divergent transcription at enhancers initiates from
transcription factor binding sites. Thus, by mapping enhancers with
modified activity, it is possible to predict what transcription
factors are responsible for these changes and to gain insight on the
signal transduction pathways involved.  Sites of divergent
transcription are best mapped using run-on data which contain only
reads from sites of transcription initiation. Yet, it has not been
possible so far to generate run-on data from patient samples, because
the quantity and the quality of the RNA yielded by the biopsies are
insufficient to generate the libraries. We therefore need to extract
information on sites of divergent transcription from classical
RNA-seq. With this type of data, the task is complicated by artefacts
generated by divergent/convergent genes, and conditions to filter away
false positives must be worked out. These may include the use of DNase
hypersensitivity or Fantom5 CAGE data.  Technically, we would like to
start with a set of publically available data including high depth
RNA-seq on a tissue or cell line where enhancers have already been
mapped (either by ChIP-seq or run-ons) and then implement a pipeline
allowing to predict the position of these enhancers as reliably as
possible using only the RNA-seq data.  Once this pipeline will be
validated, we wish to use it on multiple sclerosis patient data to
estimate changes in the enhancer landscape.  In a second step, we wish
to develop a tool allowing to identify transcription factors with
binding sites enriched within a specific set of activated (or
repressed) enhancers. This tool would use the ENCODE Chip-seq and
Factorbook repository to identify transcription factors
overrepresented in the sequences covered by the enhancers.  Finally,
as multiple sclerosis is associated with increased transcription of
endogenous retroviruses (HERVs) that may have been coopted by the
human genome because of their potential enhancer activity, we wish to
explore a possible enrichment in HERVs sequences within enhancers
modified by the disease.


*** Divergent transcription identification
**** DONE [#A] 0.1
     CLOSED: [2018-01-10 Wed 16:13]
    Using divtrans pipeline

    http://localhost:8888/notebooks/hub/divtrans/code/11.1_Divtrans_Ribo-minus.ipynb

    Improvements:
    - [ ] Identify the center of divergent transcription event and
      create shortest interval between divergent reads

**** TODO [#A] 0.2
     DEADLINE: <2018-01-12 Fri>

     - [ ] Verify that overlapping read pairs are considered

     - [ ] Add all sites of divergent transcription (ie overlapping reads)

**** TODO [#B] 0.2								     

     Checking available solutions (Found in cite:Lilam2014 ):
     
     - [ ] Testing https://github.com/Kraus-Lab/active-enhancers

       *BUG* Docker image not building
       Cannot reach archive.ubuntu and other ubuntu related depos
       
     - [ ] groHMM: 

       Designed for GRO-seq but might be of use with our
       totalRNA dataset
            
** Current workflow



** Data
*** Ribominus data

#+BEGIN_SRC sh
  cd  /home/ekornobis/data/muchardt/seqherv/data_11_2017
  pyNextGen.py *fq.gz
#+END_SRC

#+RESULTS:
| FileName            |  Nreads(M) |  Nbases(G) |      Ns | MinLen | MaxLen | MeanLen | StdevLen |         MeanGC |
| Compoud_A_1_1.fq.gz | 106.917918 | 16.0376877 | 1740681 |    150 |    150 |   150.0 |      0.0 | 0.504873103247 |
| Compoud_A_1_2.fq.gz | 106.917918 | 16.0376877 | 3510740 |    150 |    150 |   150.0 |      0.0 | 0.512369123262 |
| VEUDE_A_1_1.fq.gz   | 111.018276 | 16.6527414 | 1788598 |    150 |    150 |   150.0 |      0.0 | 0.519798430666 |
| VEUDE_A_1_2.fq.gz   | 111.018276 | 16.6527414 | 3601731 |    150 |    150 |   150.0 |      0.0 |  0.52754046556 |

*** Regions of interest 
Comparing CompoundA and Veude:

Clear enhancer:
chr2:231,487,450-231,560,819
