* DIVTRANS 							   :Muchardt:
  
  We look at the implication of HERV sequences in the reactivation of
  immune related genes.

** TODO Project description					     :ATTACH:
   DEADLINE: <2018-01-05 Fri>
   :PROPERTIES:
   :Attachments: Enhancers_of_Life.pdf
   :ID:       0f6a94ab-cd00-44b4-8a39-3ed6bf8f3414
   :END:      

   Project to be presented to the Hub:

Project title: Extracting disease-associated epimutations from patient genome-wide nuclear
run-on data

Project Context And Summary :

   Multifactorial diseases, including multiple sclerosis, are caused
not by a single gene mutation, but by a combination of genetic and
environmental factors working together to induce the
symptoms. Currently, these diseases are mostly approached by an
inventory of DNA mutations and external factors at play, bringing, in
the end, limited knowledge of the true etiology of the
disease. Alternatively, an approach allowing screening patient genomes
for sites of transcriptional dis-regulation (or epimutations rather
than DNA mutations) would directly provide information on the
mechanisms and the pathways affected by the disease.  

Available approaches to identify epimutations via changes in local
chromatin states would require unrealistic patient sample size and
exorbitant costs in terms of resources and manpower. However, recent
observations have revealed that most useful information on chromatin
states can be extracted from genome-wide nuclear run-on data and
possibly total RNA-seq data. By quantifying nascent transcripts from
enhancers and promoters, it is possible to establish “landscapes”
that, in turn, allow detection of epimutations.

Over the recent years, several teams have provided quantitative data
confirming what the field had suspected for some time, namely that
there is a direct correlation between enhancer and promoter activity
and the level of non-coding RNAs (eRNAs and PROMPTs respectively) they
produce (see for example [2] [3]). This is a major progress because it
means that it is possible to extract very informative chromatin-state
maps out of relatively simple RNA sequencing approaches.

Bioinformatics tools 

Promoters and enhancers are sites of divergent transcription. We aim
at investigating the available tools like bioconductor groHMM and
active-enhancers and if necessary develop our own bioinformatic tool
to detect divergent transcription locations in GRO and RNA-seq
datasets already available. So, at the root of the process is an
initial inventory of all sites of divergent transcription (in at least
one condition), and an identification of those sites overlapping with
annotated genes (promoters) and those that do no (enhancers). Then, at
the identified sites, we will quantify promoter and enhancer activity
modified between stimulated and unstimulated T cells, and later
between patients and healthy donors.

In a second phase, we will further mine this data for information on
chromatin states and regulators at play. In particular, we will
examine differentially active enhancers and promoters for the presence
of transcription factor binding sequences with tools like RTFBSDB
(https://doi.org/10.1093/bioinformatics/btw338) or equivalent. This
will allow gaining insight in the transcription factors and the signal
transduction pathways at play. We will also examine the enhancer
sequences for the presence of endogenous viral sequences (many HERV
sequences are transcriptionally activated in patients with MS; a role
for these viral sequences in long range transcriptional regulation
would explain this activation). Finally, in the third phase, our
objective will be to combine the more successful approaches in a
unified bioinformatics tool that will be made available to the
community to extract chromatin-state information from nuclear run on
and RNA-seq data.

Data information

Data availability : Yes
Data generated in Pasteur? 
Data description : 
Data type : 

Data analysis information
Data analysis experience :No

Additional information

** Analysis
*** Divergent transcription identification
**** TODO [#A] 0.1
    Using divtrans pipeline

    http://localhost:8888/notebooks/hub/divtrans/code/11.1_Divtrans_Ribo-minus.ipynb

    Improvements:
    - [ ] Identify the center of divergent transcription event and
      create a window of 500bp centered on it

**** TODO [#B] 0.2								     

     Checking available solutions (Found in cite:Lilam2014 ):
     
     - [ ] Testing https://github.com/Kraus-Lab/active-enhancers

       *BUG* Docker image not building
       Cannot reach archive.ubuntu and other ubuntu related depos
       
     - [ ] groHMM: 

       Designed for GRO-seq but might be of use with our
       totalRNA dataset
            
** Current workflow
   
** Data
*** Regions of interest 
Comparing CompoundA and Veude:

Clear enhancer:
chr2:231,487,450-231,560,819
