{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "extensions": {
     "jupyter_dashboards": {
      "version": 1,
      "views": {
       "grid_default": {
        "col": 0,
        "height": 4,
        "hidden": false,
        "row": 0,
        "width": 4
       },
       "report_default": {
        "hidden": false
       }
      }
     }
    }
   },
   "source": [
    "# Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-01-12T13:27:24.140077Z",
     "start_time": "2018-01-12T13:27:23.511971Z"
    },
    "collapsed": true,
    "extensions": {
     "jupyter_dashboards": {
      "version": 1,
      "views": {
       "grid_default": {
        "hidden": true
       },
       "report_default": {
        "hidden": true
       }
      }
     }
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import pandas as pd\n",
    "#from multiprocessing import Pool\n",
    "\n",
    "import bed_lib as bl\n",
    "import div_trans_lib as dt\n",
    "import basics_counts as bc\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-01-12T13:27:25.055283Z",
     "start_time": "2018-01-12T13:27:25.010690Z"
    },
    "collapsed": true,
    "extensions": {
     "jupyter_dashboards": {
      "version": 1,
      "views": {
       "grid_default": {
        "hidden": true
       },
       "report_default": {
        "hidden": true
       }
      }
     }
    }
   },
   "outputs": [],
   "source": [
    "# To not have to restart the kernel everytime changes are made in modules\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "extensions": {
     "jupyter_dashboards": {
      "version": 1,
      "views": {
       "grid_default": {
        "col": 0,
        "height": 5,
        "hidden": false,
        "row": 86,
        "width": 12
       },
       "report_default": {
        "hidden": false
       }
      }
     }
    }
   },
   "source": [
    "**WARNINGS**\n",
    "\n",
    "- TRY filtering only uniquely mapping reads (MAPQ 255) and compare with filtering only primary \n",
    "\n",
    "**TO EXPLAIN**\n",
    "\n",
    "There is a rationale being only using reads (and not fragments), since we are only interested in ends of - reads first and then beginnings of + reads \n",
    "> Actually might be interesting to try using only one reads of the pair (TO CHECK)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "extensions": {
     "jupyter_dashboards": {
      "version": 1,
      "views": {
       "grid_default": {
        "col": 7,
        "height": 4,
        "hidden": false,
        "row": 43,
        "width": 4
       },
       "report_default": {
        "hidden": false
       }
      }
     }
    }
   },
   "source": [
    "## Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-01-12T13:27:26.846959Z",
     "start_time": "2018-01-12T13:27:26.810475Z"
    },
    "extensions": {
     "jupyter_dashboards": {
      "version": 1,
      "views": {
       "grid_default": {
        "hidden": true
       },
       "report_default": {
        "hidden": true
       }
      }
     }
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/bams/VEUDE_A_1_coordSort_filt.bam',\n",
       " '../data/bams/Compoud_A_1_coordSort_filt.bam']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bams = glob.glob('../data/bams/*_coordSort_filt.bam')\n",
    "#bams = glob.glob('../data/bams/test/*.bam')\n",
    "bams"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyse from bams\n",
    "\n",
    "The idea here is to detect divergent transcription directly from a sorted bam.\n",
    "This would fix some limitations of the first method.\n",
    "\n",
    "ALGO: \n",
    "Look through the coordinated sorted bam file, registering the transcribed strand for each reads.\n",
    "\n",
    "Since we have paired strand-specific libraries (rf):\n",
    "- first of the pair, reverse complemented = + strand\n",
    "- second of the pair, NOT reverse complemented = + strand\n",
    "- first of the pair, NOT reverse complemented = - strand\n",
    "- second of the pair, reverse complemented = - strand\n",
    "\n",
    "When a change of sign is detected, calculate the distance with the previous transcript on the opposite strand \n",
    "\n",
    "## Detect divergent transcription events accross sorted bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "div_trans_beds = [dt.get_divergent_transcription_beds(bam, distance=500) for bam in bams]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\t14627\t14777\tdiv_event_1\t0\t.\n",
      "1\t14627\t14777\tdiv_event_2\t0\t.\n",
      "1\t14632\t14781\tdiv_event_3\t0\t.\n",
      "1\t14660\t14810\tdiv_event_4\t0\t.\n",
      "1\t14661\t14811\tdiv_event_5\t0\t.\n",
      "1\t14662\t14812\tdiv_event_6\t0\t.\n",
      "1\t14675\t14824\tdiv_event_7\t0\t.\n",
      "1\t14675\t14825\tdiv_event_8\t0\t.\n",
      "1\t15143\t15290\tdiv_event_9\t0\t.\n",
      "1\t15159\t15308\tdiv_event_10\t0\t.\n"
     ]
    }
   ],
   "source": [
    "div_trans_beds[0].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge the divergent transcription events detected:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter based on featureCounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "div_trans_beds_count_filtered = []\n",
    "\n",
    "for bed, bam in zip(div_trans_beds, bams):\n",
    "    div_trans_beds_count_filtered.append(dt.filter_by_counts(bed, bam, \n",
    "                                                       '../data/genome_chro_size.tab', \n",
    "                                                       flank=500, count_thres=10, count_ratio_thres=5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter based on deeptools bamCoverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using already available bed file: ../data/bams/VEUDE_A_1_coordSort_filt_forward.bed\n",
      "Using already available bed file: ../data/bams/VEUDE_A_1_coordSort_filt_reverse.bed\n"
     ]
    }
   ],
   "source": [
    "div_trans_beds_cov_filtered = {}\n",
    "\n",
    "for bed, bam in zip(div_trans_beds, bams):\n",
    "    bed_sel, df_coverages = dt.filter_by_coverage(bed, bam, '../data/genome_chro_size.tab', \n",
    "                                                  flank=500, cov_thres=0.18, cov_ratio_thres=1.9, threads=4)\n",
    "    div_trans_beds_cov_filtered[os.path.basename(bed.path)] = {'bed': bed_sel, 'coverage': df_coverages}\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14\t75725764\t75726058\tdiv_event_1\t0\t.\n",
      "14\t75745088\t75745300\tdiv_event_10\t0\t.\n",
      "14\t75760895\t75761121\tdiv_event_236\t0\t.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "div_trans_beds_cov_filtered['dieldrin_fos_area_divtrans.bed']['bed'].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "div_trans_beds_filter_merged = [div_trans_beds_cov_filtered[x]['bed'].merge() for x in div_trans_beds_cov_filtered]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14\t75725764\t75726058\n",
      "14\t75745088\t75745300\n",
      "14\t75760895\t75761121\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "div_trans_beds_filter_merged[0].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter divergent intervals events by overlap with CAGE TSS\n",
    "\n",
    "Use FANTOM5 transcription initiation information obtained with CAGE data (either simple or divergent transcription) to filter divergent transcription events detected in RNA-seq alignments.\n",
    "\n",
    "## Download cage peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 2: jvajoi: command not found\n",
      "--2018-07-20 15:12:38--  http://fantom.gsc.riken.jp/5/datafiles/latest/extra/CAGE_peaks/hg19.cage_peak_phase1and2combined_coord.bed.gz\n",
      "Resolving fantom.gsc.riken.jp... 134.160.84.66\n",
      "Connecting to fantom.gsc.riken.jp|134.160.84.66|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4405584 (4,2M) [application/x-gzip]\n",
      "Saving to: ‘../data/cage_peaks.bed.gz’\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  1% 96,3K 44s\n",
      "    50K .......... .......... .......... .......... ..........  2%  199K 32s\n",
      "   100K .......... .......... .......... .......... ..........  3% 5,98M 22s\n",
      "   150K .......... .......... .......... .......... ..........  4% 16,6M 16s\n",
      "   200K .......... .......... .......... .......... ..........  5%  203K 17s\n",
      "   250K .......... .......... .......... .......... ..........  6% 7,63M 14s\n",
      "   300K .......... .......... .......... .......... ..........  8% 10,7M 12s\n",
      "   350K .......... .......... .......... .......... ..........  9%  211K 12s\n",
      "   400K .......... .......... .......... .......... .......... 10% 5,98M 11s\n",
      "   450K .......... .......... .......... .......... .......... 11% 7,18M 10s\n",
      "   500K .......... .......... .......... .......... .......... 12% 11,6M 9s\n",
      "   550K .......... .......... .......... .......... .......... 13%  212K 9s\n",
      "   600K .......... .......... .......... .......... .......... 15% 6,40M 9s\n",
      "   650K .......... .......... .......... .......... .......... 16% 4,50M 8s\n",
      "   700K .......... .......... .......... .......... .......... 17% 6,80M 7s\n",
      "   750K .......... .......... .......... .......... .......... 18%  220K 8s\n",
      "   800K .......... .......... .......... .......... .......... 19% 16,0M 7s\n",
      "   850K .......... .......... .......... .......... .......... 20% 3,10M 7s\n",
      "   900K .......... .......... .......... .......... .......... 22% 6,12M 6s\n",
      "   950K .......... .......... .......... .......... .......... 23%  227K 7s\n",
      "  1000K .......... .......... .......... .......... .......... 24% 5,55M 6s\n",
      "  1050K .......... .......... .......... .......... .......... 25% 3,43M 6s\n",
      "  1100K .......... .......... .......... .......... .......... 26% 6,37M 6s\n",
      "  1150K .......... .......... .......... .......... .......... 27% 7,66M 5s\n",
      "  1200K .......... .......... .......... .......... .......... 29%  223K 6s\n",
      "  1250K .......... .......... .......... .......... .......... 30% 4,18M 5s\n",
      "  1300K .......... .......... .......... .......... .......... 31% 4,86M 5s\n",
      "  1350K .......... .......... .......... .......... .......... 32% 5,01M 5s\n",
      "  1400K .......... .......... .......... .......... .......... 33%  233K 5s\n",
      "  1450K .......... .......... .......... .......... .......... 34% 5,63M 5s\n",
      "  1500K .......... .......... .......... .......... .......... 36% 3,75M 5s\n",
      "  1550K .......... .......... .......... .......... .......... 37% 3,91M 4s\n",
      "  1600K .......... .......... .......... .......... .......... 38% 6,31M 4s\n",
      "  1650K .......... .......... .......... .......... .......... 39%  231K 4s\n",
      "  1700K .......... .......... .......... .......... .......... 40% 4,58M 4s\n",
      "  1750K .......... .......... .......... .......... .......... 41% 9,15M 4s\n",
      "  1800K .......... .......... .......... .......... .......... 42% 2,83M 4s\n",
      "  1850K .......... .......... .......... .......... .......... 44% 15,1M 4s\n",
      "  1900K .......... .......... .......... .......... .......... 45%  231K 4s\n",
      "  1950K .......... .......... .......... .......... .......... 46% 5,08M 4s\n",
      "  2000K .......... .......... .......... .......... .......... 47% 6,25M 3s\n",
      "  2050K .......... .......... .......... .......... .......... 48% 2,89M 3s\n",
      "  2100K .......... .......... .......... .......... .......... 49%  255K 3s\n",
      "  2150K .......... .......... .......... .......... .......... 51% 2,13M 3s\n",
      "  2200K .......... .......... .......... .......... .......... 52% 7,46M 3s\n",
      "  2250K .......... .......... .......... .......... .......... 53% 6,51M 3s\n",
      "  2300K .......... .......... .......... .......... .......... 54% 2,25M 3s\n",
      "  2350K .......... .......... .......... .......... .......... 55% 8,37M 3s\n",
      "  2400K .......... .......... .......... .......... .......... 56%  242K 3s\n",
      "  2450K .......... .......... .......... .......... .......... 58% 3,85M 3s\n",
      "  2500K .......... .......... .......... .......... .......... 59% 5,76M 2s\n",
      "  2550K .......... .......... .......... .......... .......... 60% 2,48M 2s\n",
      "  2600K .......... .......... .......... .......... .......... 61% 8,04M 2s\n",
      "  2650K .......... .......... .......... .......... .......... 62%  254K 2s\n",
      "  2700K .......... .......... .......... .......... .......... 63% 2,04M 2s\n",
      "  2750K .......... .......... .......... .......... .......... 65% 14,9M 2s\n",
      "  2800K .......... .......... .......... .......... .......... 66% 9,98M 2s\n",
      "  2850K .......... .......... .......... .......... .......... 67% 2,20M 2s\n",
      "  2900K .......... .......... .......... .......... .......... 68%  269K 2s\n",
      "  2950K .......... .......... .......... .......... .......... 69% 2,20M 2s\n",
      "  3000K .......... .......... .......... .......... .......... 70% 3,10M 2s\n",
      "  3050K .......... .......... .......... .......... .......... 72% 5,93M 2s\n",
      "  3100K .......... .......... .......... .......... .......... 73% 3,06M 1s\n",
      "  3150K .......... .......... .......... .......... .......... 74% 8,26M 1s\n",
      "  3200K .......... .......... .......... .......... .......... 75%  261K 1s\n",
      "  3250K .......... .......... .......... .......... .......... 76% 1,65M 1s\n",
      "  3300K .......... .......... .......... .......... .......... 77% 7,46M 1s\n",
      "  3350K .......... .......... .......... .......... .......... 79% 9,75M 1s\n",
      "  3400K .......... .......... .......... .......... .......... 80% 3,36M 1s\n",
      "  3450K .......... .......... .......... .......... .......... 81% 8,21M 1s\n",
      "  3500K .......... .......... .......... .......... .......... 82%  262K 1s\n",
      "  3550K .......... .......... .......... .......... .......... 83% 1,53M 1s\n",
      "  3600K .......... .......... .......... .......... .......... 84% 6,71M 1s\n",
      "  3650K .......... .......... .......... .......... .......... 85% 12,1M 1s\n",
      "  3700K .......... .......... .......... .......... .......... 87% 3,99M 1s\n",
      "  3750K .......... .......... .......... .......... .......... 88% 7,04M 1s\n",
      "  3800K .......... .......... .......... .......... .......... 89%  264K 1s\n",
      "  3850K .......... .......... .......... .......... .......... 90% 1,52M 0s\n",
      "  3900K .......... .......... .......... .......... .......... 91% 4,24M 0s\n",
      "  3950K .......... .......... .......... .......... .......... 92% 21,3M 0s\n",
      "  4000K .......... .......... .......... .......... .......... 94% 5,23M 0s\n",
      "  4050K .......... .......... .......... .......... .......... 95% 4,10M 0s\n",
      "  4100K .......... .......... .......... .......... .......... 96%  270K 0s\n",
      "  4150K .......... .......... .......... .......... .......... 97% 1,60M 0s\n",
      "  4200K .......... .......... .......... .......... .......... 98% 2,69M 0s\n",
      "  4250K .......... .......... .......... .......... .......... 99% 8,85M 0s\n",
      "  4300K ..                                                    100% 4441G=5,1s\n",
      "\n",
      "2018-07-20 15:12:43 (842 KB/s) - ‘../data/cage_peaks.bed.gz’ saved [4405584/4405584]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "jvajoi\n",
    "\n",
    "wget http://fantom.gsc.riken.jp/5/datafiles/latest/extra/CAGE_peaks/hg19.cage_peak_phase1and2combined_coord.bed.gz \\\n",
    "    -O ../data/cage_peaks.bed.gz\n",
    "    \n",
    "gunzip ../data/cage_peaks.bed.gz\n",
    "    \n",
    "# Convert to Ensembl type bed\n",
    "sed 's/^chr//g' ../data/cage_peaks.bed | bedtools sort -i - > ../data/cage_peaks_clean.bed\n",
    "\n",
    "# Separate two beds (one per strand) to later catch both cage transcription direction\n",
    "#awk '$6~\"+\" {print}' ../data/cage_peaks_clean.bed > ../data/cage_peaks_plusStrand.bed\n",
    "#awk '$6~\"-\" {print}' ../data/cage_peaks_clean.bed > ../data/cage_peaks_minusStrand.bed\n",
    "\n",
    "# Cleanup\n",
    "rm ../data/cage_peaks.bed\n",
    "#rm ../data/cage_peaks_clean.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter divergent transcription events with CAGE dataset\n",
    "### Detect divergent transcription in CAGE dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "cage_bed = bl.Bed('../data/cage_peaks_clean.bed')\n",
    "cage_div_trans_out = '../data/cage_div_trans.bed'\n",
    "\n",
    "with open(cage_div_trans_out, 'w') as f:\n",
    "        for interval in dt.identify_divergent_transcription(cage_bed.get_intervals()):\n",
    "            f.write('\\t'.join([str(x) for x in interval]) + '\\n')\n",
    "        \n",
    "cage_div_trans_bed = bl.Bed(cage_div_trans_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select RNA-seq divtrans events overlapping with CAGE divtrans events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "div_trans_intersect_cage = [bed.intersect(cage_div_trans_bed, supp_args='-wa')\n",
    "                                for bed in div_trans_beds_filter_merged]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Nbases</th>\n",
       "      <th>Nintervals</th>\n",
       "      <th>bedobj</th>\n",
       "      <th>name</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;Bed object: dieldrin_fos_area_divtrans_covera...</td>\n",
       "      <td>dieldrin_fos_area_divtrans_coverage_filtered_M...</td>\n",
       "      <td>bed_outfolder/dieldrin_fos_area_divtrans_cover...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;Bed object: vehicule_fos_area_divtrans_covera...</td>\n",
       "      <td>vehicule_fos_area_divtrans_coverage_filtered_M...</td>\n",
       "      <td>bed_outfolder/vehicule_fos_area_divtrans_cover...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Nbases  Nintervals                                             bedobj  \\\n",
       "0       0           0  <Bed object: dieldrin_fos_area_divtrans_covera...   \n",
       "1       0           0  <Bed object: vehicule_fos_area_divtrans_covera...   \n",
       "\n",
       "                                                name  \\\n",
       "0  dieldrin_fos_area_divtrans_coverage_filtered_M...   \n",
       "1  vehicule_fos_area_divtrans_coverage_filtered_M...   \n",
       "\n",
       "                                                path  \n",
       "0  bed_outfolder/dieldrin_fos_area_divtrans_cover...  \n",
       "1  bed_outfolder/vehicule_fos_area_divtrans_cover...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame([bed.stats() for bed in div_trans_intersect_cage])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Nbases</th>\n",
       "      <th>Nintervals</th>\n",
       "      <th>bedobj</th>\n",
       "      <th>name</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>732</td>\n",
       "      <td>3</td>\n",
       "      <td>&lt;Bed object: dieldrin_fos_area_divtrans_covera...</td>\n",
       "      <td>dieldrin_fos_area_divtrans_coverage_filtered_M</td>\n",
       "      <td>bed_outfolder/dieldrin_fos_area_divtrans_cover...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;Bed object: vehicule_fos_area_divtrans_covera...</td>\n",
       "      <td>vehicule_fos_area_divtrans_coverage_filtered_M</td>\n",
       "      <td>bed_outfolder/vehicule_fos_area_divtrans_cover...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Nbases  Nintervals                                             bedobj  \\\n",
       "0     732           3  <Bed object: dieldrin_fos_area_divtrans_covera...   \n",
       "1       0           0  <Bed object: vehicule_fos_area_divtrans_covera...   \n",
       "\n",
       "                                             name  \\\n",
       "0  dieldrin_fos_area_divtrans_coverage_filtered_M   \n",
       "1  vehicule_fos_area_divtrans_coverage_filtered_M   \n",
       "\n",
       "                                                path  \n",
       "0  bed_outfolder/dieldrin_fos_area_divtrans_cover...  \n",
       "1  bed_outfolder/vehicule_fos_area_divtrans_cover...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame([bed.stats() for bed in div_trans_beds_filter_merged])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare to FANTOM5 CAGE enhancer prediction [SLIDEBASE]\n",
    "\n",
    "One can use SLIDEBASE for enhancer http://slidebase.binf.ku.dk/human_enhancers/results\n",
    "\n",
    "Downloading bed from SLIDEBASE (same as SLIDEBASE \"Download BED\"):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2018-07-20 15:17:29--  http://slidebase.binf.ku.dk/human_enhancers/bed\n",
      "Resolving slidebase.binf.ku.dk... 130.226.13.22\n",
      "Connecting to slidebase.binf.ku.dk|130.226.13.22|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: unspecified [text/plain]\n",
      "Saving to: ‘../data/slidebase_enhancers.bed’\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  752K\n",
      "    50K .......... .......... .......... .......... .......... 1,82M\n",
      "   100K .......... .......... .......... .......... .......... 3,57M\n",
      "   150K .......... .......... .......... .......... .......... 8,45M\n",
      "   200K .......... .......... .......... .......... .......... 2,27M\n",
      "   250K .......... .......... .......... .......... .......... 5,53M\n",
      "   300K .......... .......... .......... .......... .......... 11,4M\n",
      "   350K .......... .......... .......... .......... .......... 10,8M\n",
      "   400K .......... .......... .......... .......... .......... 4,23M\n",
      "   450K .......... .......... .......... .......... .......... 9,43M\n",
      "   500K .......... .......... .......... .......... .......... 10,6M\n",
      "   550K .......... .......... .......... .......... .......... 14,1M\n",
      "   600K .......... .......... .......... .......... .......... 10,5M\n",
      "   650K .......... .......... .......... .......... .......... 15,0M\n",
      "   700K .......... .......... .......... .......... ..........  586K\n",
      "   750K .......... .......... .......... .......... .......... 11,7M\n",
      "   800K .......... .......... .......... .......... .......... 11,0M\n",
      "   850K .......... .......... .......... .......... .......... 6,93M\n",
      "   900K .......... .......... .......... .......... .......... 15,9M\n",
      "   950K .......... .......... .......... .......... .......... 7,55M\n",
      "  1000K .......... .......... .......... .......... .......... 10,1M\n",
      "  1050K .......... .......... .......... .......... .......... 12,2M\n",
      "  1100K .......... .......... .......... .......... .......... 14,4M\n",
      "  1150K .......... .......... .......... .......... .......... 6,74M\n",
      "  1200K .......... .......... .......... .......... .......... 16,9M\n",
      "  1250K .......... .......... .......... .......... .......... 7,98M\n",
      "  1300K .......... .......... .......... .......... .......... 12,3M\n",
      "  1350K .......... .......... .......... .......... .......... 14,8M\n",
      "  1400K .......... .......... .......... .......... .......... 9,63M\n",
      "  1450K .......... .......... .......... .......... .......... 12,4M\n",
      "  1500K .......... .......... .......... .......... ..........  800K\n",
      "  1550K .......... .......... .......... .......... .......... 4,73M\n",
      "  1600K .......... .......... .......... .......... .......... 1,03M\n",
      "  1650K ....                                                   8620G=0,5s\n",
      "\n",
      "2018-07-20 15:17:29 (3,51 MB/s) - ‘../data/slidebase_enhancers.bed’ saved [1694228]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "    \n",
    "wget http://slidebase.binf.ku.dk/human_enhancers/bed -O ../data/slidebase_enhancers.bed\n",
    "\n",
    "# Convert to Ensembl type bed\n",
    "sed 's/^chr//g' ../data/slidebase_enhancers.bed | bedtools sort -i - > ../data/slidebase_enhancers_clean.bed\n",
    "\n",
    "rm ../data/slidebase_enhancers.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "slidebase_enh = bl.Bed('../data/slidebase_enhancers_clean.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the chromosome sizes for statistical testing of interval association:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seq_names\n",
       "1             249250621\n",
       "10            135534747\n",
       "11            135006516\n",
       "12            133851895\n",
       "13            115169878\n",
       "14            107349540\n",
       "15            102531392\n",
       "16             90354753\n",
       "17             81195210\n",
       "18             78077248\n",
       "19             59128983\n",
       "2             243199373\n",
       "20             63025520\n",
       "21             48129895\n",
       "22             51304566\n",
       "3             198022430\n",
       "4             191154276\n",
       "5             180915260\n",
       "6             171115067\n",
       "7             159138663\n",
       "8             146364022\n",
       "9             141213431\n",
       "MT                16569\n",
       "X             155270560\n",
       "Y              59373566\n",
       "GL000192.1       547496\n",
       "GL000225.1       211173\n",
       "GL000194.1       191469\n",
       "GL000193.1       189789\n",
       "GL000200.1       187035\n",
       "                ...    \n",
       "GL000233.1        45941\n",
       "GL000237.1        45867\n",
       "GL000230.1        43691\n",
       "GL000242.1        43523\n",
       "GL000243.1        43341\n",
       "GL000241.1        42152\n",
       "GL000236.1        41934\n",
       "GL000240.1        41933\n",
       "GL000206.1        41001\n",
       "GL000232.1        40652\n",
       "GL000234.1        40531\n",
       "GL000202.1        40103\n",
       "GL000238.1        39939\n",
       "GL000244.1        39929\n",
       "GL000248.1        39786\n",
       "GL000196.1        38914\n",
       "GL000249.1        38502\n",
       "GL000246.1        38154\n",
       "GL000203.1        37498\n",
       "GL000197.1        37175\n",
       "GL000245.1        36651\n",
       "GL000247.1        36422\n",
       "GL000201.1        36148\n",
       "GL000235.1        34474\n",
       "GL000239.1        33824\n",
       "GL000210.1        27682\n",
       "GL000231.1        27386\n",
       "GL000229.1        19913\n",
       "GL000226.1        15008\n",
       "GL000207.1         4262\n",
       "Name: seq_len, Length: 84, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bl.get_chrom_sizes('/home/khourhin/data/genomes/Homo_sapiens.GRCh37.dna_sm.primary_assembly.fa', '../data/genome_chro_size.tab')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fisher test\n",
    "Testing if significant association between intervals detected with divergent transcription and the ones reported in fantom5 CAGE enhancer dataset:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'left': 0.99501,\n",
       "  'right': 1.0,\n",
       "  'two-tail': 1.0,\n",
       "  'ratio': 0.0,\n",
       "  'bed1': 'dieldrin_fos_area_divtrans_coverage_filtered_M',\n",
       "  'bed2': 'slidebase_enhancers_clean'},\n",
       " {'left': 1.0,\n",
       "  'right': 1.0,\n",
       "  'two-tail': 1.0,\n",
       "  'ratio': nan,\n",
       "  'bed1': 'vehicule_fos_area_divtrans_coverage_filtered_M',\n",
       "  'bed2': 'slidebase_enhancers_clean'}]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[bed.fisher(slidebase_enh, '../data/genome_chro_size.tab') for bed in div_trans_beds_filter_merged]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Jaccard index\n",
    "\n",
    "Evaluating the distance between enhancer intervals defined by fantom5 and the ones predicted from RNA-seq data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>intersection</th>\n",
       "      <th>union-intersection</th>\n",
       "      <th>jaccard</th>\n",
       "      <th>n_intersections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dieldrin_fos_area_divtrans_coverage_filtered_M</th>\n",
       "      <td>0</td>\n",
       "      <td>2867950</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vehicule_fos_area_divtrans_coverage_filtered_M</th>\n",
       "      <td>0</td>\n",
       "      <td>392</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                intersection  \\\n",
       "dieldrin_fos_area_divtrans_coverage_filtered_M             0   \n",
       "vehicule_fos_area_divtrans_coverage_filtered_M             0   \n",
       "\n",
       "                                                union-intersection  jaccard  \\\n",
       "dieldrin_fos_area_divtrans_coverage_filtered_M             2867950        0   \n",
       "vehicule_fos_area_divtrans_coverage_filtered_M                 392        0   \n",
       "\n",
       "                                                n_intersections  \n",
       "dieldrin_fos_area_divtrans_coverage_filtered_M                0  \n",
       "vehicule_fos_area_divtrans_coverage_filtered_M                0  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([bed.jaccard(slidebase_enh) for bed in div_trans_beds_filter_merged])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Shuffling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "shuffled_slidebase_enh = slidebase_enh.multi_shuffle(100, '../data/genome_chro_size.tab')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Average number of common intervals between shuffled CAGE dataset and divergent transcription inferred dataset:\n",
    "\n",
    "(to relate with n_intersection from jaccard above)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1201.34"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame([div_trans_beds_merged[0].intersect(x).stats() for x in shuffled_slidebase_enh]).Nintervals.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1290.88"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame([div_trans_beds_merged[1].intersect(x).stats() for x in shuffled_slidebase_enh]).Nintervals.mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**CCL:** ~2 times more intersections between divtrans and CAGE intervals than expected by chance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stats on divergent transcription intervals\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersection_cage_div_trans_bed = [x.intersect(slidebase_enh) for x in div_trans_beds_filter_merged]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare to Transcription factors intervals\n",
    "\n",
    "Download transcription factors bed from UCSC:\n",
    "\n",
    "**IMPORTANT NOTE:** At the difference of enhancer and div_trans intervals, transcription factors intervals are not merged \n",
    "> This leads to divergent estimations in intersection when considering jaccard and intersect results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Download from UCSC mysql database, remove the 'chr',the header, sort the bed\n",
    "mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -e \"use hg19; SELECT * FROM wgEncodeRegTfbsClusteredV3;\" | \\\n",
    "    awk -v OFS='\\t' 'NR>1 {gsub(\"chr\",\"\",$2); print $2, $3, $4, $5, $6}' | \\\n",
    "    bedtools sort -i - > ../data/wgEncodeRegTfbsClusteredV3.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcription_factors_bed = bl.Bed('../data/wgEncodeRegTfbsClusteredV3.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fisher test\n",
    "Testing if significant association between intervals detected with divergent transcription and transcription factors from UCSC\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'left': 1.0,\n",
       "  'right': 5.6893e-103,\n",
       "  'two-tail': 5.6893e-103,\n",
       "  'ratio': inf,\n",
       "  'bed1': 'dieldrin_fos_area_divtrans_coverage_filtered_M',\n",
       "  'bed2': 'wgEncodeRegTfbsClusteredV3'},\n",
       " {'left': 1.0,\n",
       "  'right': 1.0,\n",
       "  'two-tail': 1.0,\n",
       "  'ratio': nan,\n",
       "  'bed1': 'vehicule_fos_area_divtrans_coverage_filtered_M',\n",
       "  'bed2': 'wgEncodeRegTfbsClusteredV3'}]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[bed.fisher(transcription_factors_bed, '../data/genome_chro_size.tab') for bed in div_trans_beds_filter_merged]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Jaccard index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>intersection</th>\n",
       "      <th>union-intersection</th>\n",
       "      <th>jaccard</th>\n",
       "      <th>n_intersections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dieldrin_fos_area_divtrans_coverage_filtered_M</th>\n",
       "      <td>732</td>\n",
       "      <td>107553159</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vehicule_fos_area_divtrans_coverage_filtered_M</th>\n",
       "      <td>0</td>\n",
       "      <td>340</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                intersection  \\\n",
       "dieldrin_fos_area_divtrans_coverage_filtered_M           732   \n",
       "vehicule_fos_area_divtrans_coverage_filtered_M             0   \n",
       "\n",
       "                                                union-intersection   jaccard  \\\n",
       "dieldrin_fos_area_divtrans_coverage_filtered_M           107553159  0.000007   \n",
       "vehicule_fos_area_divtrans_coverage_filtered_M                 340  0.000000   \n",
       "\n",
       "                                                n_intersections  \n",
       "dieldrin_fos_area_divtrans_coverage_filtered_M                3  \n",
       "vehicule_fos_area_divtrans_coverage_filtered_M                0  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([bed.jaccard(transcription_factors_bed) for bed in div_trans_beds_filter_merged])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Shuffling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-183-3bbad0ecb528>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mshuffled_transcription_factors_beds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtranscription_factors_bed\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmulti_shuffle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'../data/genome_chro_size.tab'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/Programming/pyNextGen/bed_lib.py\u001b[0m in \u001b[0;36mmulti_shuffle\u001b[0;34m(self, nshuffle, genome_chro_size, supp_args, outfolder)\u001b[0m\n\u001b[1;32m    290\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    291\u001b[0m             \u001b[0mlogger\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Running: {}'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 292\u001b[0;31m             \u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck_output\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    293\u001b[0m             \u001b[0mshuffle_beds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mBed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mshuffle_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    294\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/programs/anaconda3/lib/python3.6/subprocess.py\u001b[0m in \u001b[0;36mcheck_output\u001b[0;34m(timeout, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    334\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    335\u001b[0m     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,\n\u001b[0;32m--> 336\u001b[0;31m                **kwargs).stdout\n\u001b[0m\u001b[1;32m    337\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    338\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/programs/anaconda3/lib/python3.6/subprocess.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(input, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    403\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mPopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mpopenargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    404\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 405\u001b[0;31m             \u001b[0mstdout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    406\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mTimeoutExpired\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    407\u001b[0m             \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkill\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/programs/anaconda3/lib/python3.6/subprocess.py\u001b[0m in \u001b[0;36mcommunicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m    828\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_stdin_write\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    829\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 830\u001b[0;31m                 \u001b[0mstdout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    831\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    832\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "shuffled_transcription_factors_beds = transcription_factors_bed.multi_shuffle(100, '../data/genome_chro_size.tab')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Average number of common intervals between shuffled transcription factors intervals and divergent transcription inferred dataset:\n",
    "\n",
    "(to relate with n_intersection from jaccard above)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame([div_trans_beds_merged[0].intersect(x).stats() for x in shuffled_transcription_factors_beds]).Nintervals.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame([div_trans_beds_merged[1].intersect(x).stats() for x in shuffled_transcription_factors_beds]).Nintervals.mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**CCL:** Only 10% of intersections of the ones expected by chance "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stats on divergent transcription intervals\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Nbases</th>\n",
       "      <th>Nintervals</th>\n",
       "      <th>bedobj</th>\n",
       "      <th>name</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>238482956</td>\n",
       "      <td>2279683</td>\n",
       "      <td>&lt;Bed object: VEUDE_A_1_coordSort_filt_divtrans...</td>\n",
       "      <td>VEUDE_A_1_coordSort_filt_divtrans-inter-wgEnco...</td>\n",
       "      <td>bed_outfolder/VEUDE_A_1_coordSort_filt_divtran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>246101331</td>\n",
       "      <td>2348583</td>\n",
       "      <td>&lt;Bed object: Compoud_A_1_coordSort_filt_divtra...</td>\n",
       "      <td>Compoud_A_1_coordSort_filt_divtrans-inter-wgEn...</td>\n",
       "      <td>bed_outfolder/Compoud_A_1_coordSort_filt_divtr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Nbases  Nintervals                                             bedobj  \\\n",
       "0  238482956     2279683  <Bed object: VEUDE_A_1_coordSort_filt_divtrans...   \n",
       "1  246101331     2348583  <Bed object: Compoud_A_1_coordSort_filt_divtra...   \n",
       "\n",
       "                                                name  \\\n",
       "0  VEUDE_A_1_coordSort_filt_divtrans-inter-wgEnco...   \n",
       "1  Compoud_A_1_coordSort_filt_divtrans-inter-wgEn...   \n",
       "\n",
       "                                                path  \n",
       "0  bed_outfolder/VEUDE_A_1_coordSort_filt_divtran...  \n",
       "1  bed_outfolder/Compoud_A_1_coordSort_filt_divtr...  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersection_transfact_div_trans_bed = [x.intersect(transcription_factors_bed) for x in div_trans_beds]\n",
    "pd.DataFrame([x.stats() for x in intersection_transfact_div_trans_bed])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing SLIDEBASE enhancer and transcription factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bed1': 'slidebase_enhancers_clean',\n",
       " 'bed2': 'wgEncodeRegTfbsClusteredV3',\n",
       " 'left': 1.0,\n",
       " 'ratio': inf,\n",
       " 'right': 0.0,\n",
       " 'two-tail': 0.0}"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slidebase_enh.fisher(transcription_factors_bed, '../data/genome_chro_size.tab')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>intersection</th>\n",
       "      <th>union-intersection</th>\n",
       "      <th>jaccard</th>\n",
       "      <th>n_intersections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>slidebase_enhancers_clean</th>\n",
       "      <td>845531</td>\n",
       "      <td>180475823</td>\n",
       "      <td>0.004685</td>\n",
       "      <td>2843</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           intersection  union-intersection   jaccard  \\\n",
       "slidebase_enhancers_clean        845531           180475823  0.004685   \n",
       "\n",
       "                           n_intersections  \n",
       "slidebase_enhancers_clean             2843  "
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slidebase_enh.jaccard(transcription_factors_bed, '../data/genome_chro_size.tab')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31134.32"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame([transcription_factors_bed.intersect(x).stats() for x in shuffled_slidebase_enh]).Nintervals.mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**CCL:** Enhancer and transcription factor intervals appears to repel each other\n",
    "> **I DON'T UNDERSTAND THIS !**\n",
    "\n",
    ">> This could be explained by transcription factors being present couple hundreds bases up or downstream (ie related but not overlapping). Might have to try augmenting div_trans intervals sizes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Length distribution of intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f05aaefcb38>"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYkAAAD8CAYAAACCRVh7AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAEylJREFUeJzt3X+MZWd93/H3p45xIhvF6zoebW1La6ptFSduHXdkLFFFQ2n8iz8MUqiMLFgD1UatrYC6/WNJpJoGITlVISoqdbTUq5iG4rgBxCps6mwdRog/DLapsb04jidmi5ddeZWaGAYk2qXf/nGfMZf1PLMzd+/O3Hv3/ZKu7rnPfc65z3fOrD8+zzn3TKoKSZJW87e2egCSpMllSEiSugwJSVKXISFJ6jIkJEldhoQkqcuQkCR1GRKSpC5DQpLU9TNbPYC1XHrppbVjx46R1v3BD37AhRdeON4BTYBZrMuapsMs1gSzWdcTTzzx11X1C+PY1kSHxI4dO3j88cdHWndxcZGFhYXxDmgCzGJd1jQdZrEmmM26kvyvcW3L6SZJUpchIUnqMiQkSV2GhCSpy5CQJHUZEpKkLkNCktRlSEiSugwJSVLXRH/jepbt2PvFV5eP3PvWLRyJJPV5JCFJ6jIkJEldhoQkqcuQkCR1nTYkklyZ5EtJnk1yOMn7W/uHknwnyZPtcevQOh9MspTkuSQ3DbXf3NqWkuw9OyVJksZlPVc3nQT2VNXXk7weeCLJofbe71XVvx/unORq4Hbgl4C/A/yPJH+vvf0J4NeAo8BjSQ5U1TfHUYgkafxOGxJVdRw43pa/n+RZ4PI1VrkNeLCqfgR8K8kScH17b6mqXgBI8mDra0hI0oRKVa2/c7ID+DLwy8C/Au4Evgc8zuBo47tJ/iPwaFX9YVvnfuBP2yZurqp/3trfBbyxqu4+5TN2A7sB5ubm/tGDDz44UmHLy8tcdNFFI627GZ7+ziuvLl9z+c+ve71Jr2sU1jQdZrEmmM263vzmNz9RVfPj2Na6v0yX5CLgs8AHqup7Se4DPgxUe/4o8F4gq6xerH7+4zUJVVX7gH0A8/PzNeqfFZz0P0l45/CX6e5YWPd6k17XKKxpOsxiTTC7dY3LukIiyfkMAuLTVfU5gKp6aej9TwJ/0l4eBa4cWv0K4Fhb7rWfE4a/ZS1J02A9VzcFuB94tqo+NtS+fajb24Fn2vIB4PYkFyS5CtgJfA14DNiZ5Kokr2NwcvvAeMqQJJ0N6zmSeBPwLuDpJE+2tt8C3pnkWgZTRkeA3wCoqsNJHmJwQvokcFdV/Rggyd3Aw8B5wP6qOjzGWiRJY7aeq5u+wurnGQ6usc5HgI+s0n5wrfUkSZPFu8CeBd7hVdKs8LYckqQujyTOgEcMkmadIbFBXsYq6VzidJMkqcuQkCR1GRKSpC7PSayD5yEknasMiTExSCTNIqebJEldhoQkqcvpprPMaShJ08wjCUlSlyEhSeoyJCRJXYaEJKnLkJAkdRkSkqQuL4Ht8NJVSfJIQpK0BkNCktRlSEiSugwJSVKXISFJ6jIkJEldhoQkqcuQkCR1GRKSpC5DQpLUZUhIkrpOGxJJrkzypSTPJjmc5P2t/ZIkh5I83563tfYk+XiSpSRPJbluaFu7Wv/nk+w6e2VJksZhPUcSJ4E9VfWLwA3AXUmuBvYCj1TVTuCR9hrgFmBne+wG7oNBqAD3AG8ErgfuWQkWSdJkOm1IVNXxqvp6W/4+8CxwOXAb8EDr9gDwtrZ8G/CpGngUuDjJduAm4FBVvVxV3wUOATePtRpJ0lht6JxEkh3ArwBfBeaq6jgMggS4rHW7HHhxaLWjra3XLkmaUOv+exJJLgI+C3ygqr6XpNt1lbZao/3Uz9nNYJqKubk5FhcX1zvEn7K8vDzyugB7rjk58robtZFxnmldk8iapsMs1gSzW9e4rCskkpzPICA+XVWfa80vJdleVcfbdNKJ1n4UuHJo9SuAY6194ZT2xVM/q6r2AfsA5ufna2Fh4dQu67K4uMio6wLcuYl/dOjIHQvr7numdU0ia5oOs1gTzG5d47Keq5sC3A88W1UfG3rrALByhdIu4AtD7e9uVzndALzSpqMeBm5Msq2dsL6xtUmSJtR6jiTeBLwLeDrJk63tt4B7gYeSvA/4NvCO9t5B4FZgCfgh8B6Aqno5yYeBx1q/36mql8dShSTprDhtSFTVV1j9fALAW1bpX8BdnW3tB/ZvZICSpK3jN64lSV2GhCSpy5CQJHUZEpKkLkNCktRlSEiSugwJSVKXISFJ6lr3Df7OBTs28X5NkjQNPJKQJHUZEpKkLkNCktRlSEiSugwJSVKXISFJ6jIkJEldhoQkqcuQkCR1GRKSpC5DQpLUZUhIkroMCUlSlyEhSeoyJCRJXYaEJKnLkJAkdRkSkqQuQ0KS1GVISJK6DAlJUpchIUnqOm1IJNmf5ESSZ4baPpTkO0mebI9bh977YJKlJM8luWmo/ebWtpRk7/hLkSSN23qOJP4AuHmV9t+rqmvb4yBAkquB24Ffauv8pyTnJTkP+ARwC3A18M7WV5I0wX7mdB2q6stJdqxze7cBD1bVj4BvJVkCrm/vLVXVCwBJHmx9v7nhEUuSNs2ZnJO4O8lTbTpqW2u7HHhxqM/R1tZrlyRNsNMeSXTcB3wYqPb8UeC9QFbpW6weRrXahpPsBnYDzM3Nsbi4ONIAl5eXN7zunmtOjvRZZ2oj4xylrklnTdNhFmuC2a1rXEYKiap6aWU5ySeBP2kvjwJXDnW9AjjWlnvtp257H7APYH5+vhYWFkYZIouLi2x03Tv3fnGkzzpTR+5YWHffUeqadNY0HWaxJpjdusZlpOmmJNuHXr4dWLny6QBwe5ILklwF7AS+BjwG7ExyVZLXMTi5fWD0YUuSNsNpjySSfAZYAC5NchS4B1hIci2DKaMjwG8AVNXhJA8xOCF9Erirqn7ctnM38DBwHrC/qg6PvRpJ0lit5+qmd67SfP8a/T8CfGSV9oPAwQ2NTpK0pfzGtSSpy5CQJHUZEpKkLkNCktRlSEiSugwJSVKXISFJ6hr13k0aox1DtwM5cu9bt3AkkvTTPJKQJHUZEpKkLkNCktRlSEiSugwJSVKXISFJ6jIkJEldhoQkqcuQkCR1GRKSpC5DQpLUZUhIkroMCUlSlyEhSeoyJCRJXYaEJKnLkJAkdZ3zf5lu+K/CSZJ+mkcSkqQuQ0KS1GVISJK6DAlJUpchIUnqOm1IJNmf5ESSZ4baLklyKMnz7Xlba0+SjydZSvJUkuuG1tnV+j+fZNfZKWf67dj7xVcfkrTV1nMk8QfAzae07QUeqaqdwCPtNcAtwM722A3cB4NQAe4B3ghcD9yzEiySpMl12pCoqi8DL5/SfBvwQFt+AHjbUPunauBR4OIk24GbgENV9XJVfRc4xGuDR5I0YUb9Mt1cVR0HqKrjSS5r7ZcDLw71O9raeu2vkWQ3g6MQ5ubmWFxcHGmAy8vL61p3zzUnR9r+Zlht/Outa5pY03SYxZpgdusal3F/4zqrtNUa7a9trNoH7AOYn5+vhYWFkQayuLjIeta9c4Ln/o/csfCatvXWNU2saTrMYk0wu3WNy6hXN73UppFozyda+1HgyqF+VwDH1miXJE2wUUPiALByhdIu4AtD7e9uVzndALzSpqUeBm5Msq2dsL6xtUmSJthpp5uSfAZYAC5NcpTBVUr3Ag8leR/wbeAdrftB4FZgCfgh8B6Aqno5yYeBx1q/36mqU0+GS5ImzGlDoqre2XnrLav0LeCuznb2A/s3NDpJ0pbyG9eSpC5DQpLUZUhIkroMCUlSlyEhSeoyJCRJXYaEJKnLkJAkdRkSkqQuQ0KS1GVISJK6DAlJUpchIUnqGvdfptMY7Rj6q3lH7n3rFo5E0rnKIwlJUpchIUnqMiQkSV2GhCSpy5CQJHUZEpKkLkNCktRlSEiSugwJSVLXTH/j2m8sS9KZmemQmCUrgbfnmpMsbO1QJJ1DnG6SJHWdk0cSw9NQkqQ+jyQkSV2GhCSpy5CQJHWdUUgkOZLk6SRPJnm8tV2S5FCS59vzttaeJB9PspTkqSTXjaMASdLZM44jiTdX1bVVNd9e7wUeqaqdwCPtNcAtwM722A3cN4bPliSdRWdjuuk24IG2/ADwtqH2T9XAo8DFSbafhc+XJI3JmYZEAX+W5Ikku1vbXFUdB2jPl7X2y4EXh9Y92tokSRPqTL8n8aaqOpbkMuBQkr9Yo29WaavXdBqEzW6Aubk5FhcXRxrY8vIye6758auvh7ez55qTI21zEsz9HCP/TCbV8vKyNU2BWawJZreucTmjkKiqY+35RJLPA9cDLyXZXlXH23TSidb9KHDl0OpXAMdW2eY+YB/A/Px8LSwsjDS2xcVFPvqVH7z6+sgdP9nOnVP8Zbo915zkn434M5lUi4uLjLqfJ5U1TY9ZrWtcRp5uSnJhktevLAM3As8AB4Bdrdsu4Att+QDw7naV0w3AKyvTUpKkyXQmRxJzwOeTrGznv1bVf0/yGPBQkvcB3wbe0fofBG4FloAfAu85g88+p3l3W0mbZeSQqKoXgH+4Svv/Bt6ySnsBd436eWfK+zVJ0sb5jWtJUpchIUnqOidvFT5LTp1G8xyFpHHySEKS1GVISJK6DAlJUpchIUnqMiQkSV2GhCSpy5CQJHX5PYkZ432dJI2TRxKSpC5DQpLUZUhIkroMCUlSlyEhSery6qYZ1vtDS171JGm9PJKQJHUZEpKkLkNCktRlSEiSujxxfQ7y1h2S1suQOMcZGJLW4nSTJKnLkJAkdTndpFc59STpVB5JSJK6PJLQqrylhyTwSEKStAaPJLQhHmFI5xaPJCRJXZt+JJHkZuA/AOcB/7mq7t3sMWj8PMKQZtOmhkSS84BPAL8GHAUeS3Kgqr65mePQ5vGyWmm6bfaRxPXAUlW9AJDkQeA2wJA4B/SONvZcc5I723sGiTRZNjskLgdeHHp9FHjjJo9BE6wXJJNiOMScYtO5YLNDIqu01U91SHYDu9vL5STPjfhZlwJ/PeK6E+s3Z7Cuaaopv7vuPlNT0wbMYk0wm3X9/XFtaLND4ihw5dDrK4Bjwx2qah+w70w/KMnjVTV/ptuZNLNYlzVNh1msCWazriSPj2tbm30J7GPAziRXJXkdcDtwYJPHIElap009kqiqk0nuBh5mcAns/qo6vJljkCSt36Z/T6KqDgIHN+GjznjKakLNYl3WNB1msSaYzbrGVlOq6vS9JEnnJG/LIUnqmsmQSHJzkueSLCXZu9Xj2YgkR5I8neTJlSsUklyS5FCS59vzttaeJB9vdT6V5LqtHf1PJNmf5ESSZ4baNlxHkl2t//NJdm1FLUNjWa2mDyX5TttfTya5dei9D7aankty01D7xPx+JrkyyZeSPJvkcJL3t/ap3Vdr1DS1+yrJzyb5WpJvtJr+bWu/KslX28/8j9oFQSS5oL1eau/vGNrWqrV2VdVMPRicEP8r4A3A64BvAFdv9bg2MP4jwKWntP07YG9b3gv8blu+FfhTBt8/uQH46laPf2jMvwpcBzwzah3AJcAL7XlbW942YTV9CPjXq/S9uv3uXQBc1X4nz5u0309gO3BdW3498Jdt7FO7r9aoaWr3Vft5X9SWzwe+2n7+DwG3t/bfB/5FW/6XwO+35duBP1qr1rU+exaPJF699UdV/R9g5dYf0+w24IG2/ADwtqH2T9XAo8DFSbZvxQBPVVVfBl4+pXmjddwEHKqql6vqu8Ah4OazP/rVdWrquQ14sKp+VFXfApYY/G5O1O9nVR2vqq+35e8DzzK4M8LU7qs1auqZ+H3Vft7L7eX57VHAPwH+uLWfup9W9t8fA29JEvq1ds1iSKx264+1fkEmTQF/luSJDL59DjBXVcdh8A8AuKy1T1utG61jWuq7u0297F+ZlmEKa2pTEr/C4P9SZ2JfnVITTPG+SnJekieBEwxC+K+Av6mqk6uM79Wxt/dfAf42I9Q0iyFx2lt/TLg3VdV1wC3AXUl+dY2+017ril4d01DffcDfBa4FjgMfbe1TVVOSi4DPAh+oqu+t1XWVtomsa5WapnpfVdWPq+paBnequB74xdW6teex1TSLIXHaW39Msqo61p5PAJ9n8Mvw0so0Uns+0bpPW60brWPi66uql9o/3v8HfJKfHLpPTU1JzmfwH9NPV9XnWvNU76vVapqFfQVQVX8DLDI4J3FxkpXvuw2P79Wxt/d/nsFU6YZrmsWQmNpbfyS5MMnrV5aBG4FnGIx/5WqRXcAX2vIB4N3tipMbgFdWpggm1EbreBi4Mcm2NjVwY2ubGKecA3o7g/0Fg5pub1eZXAXsBL7GhP1+tnnq+4Fnq+pjQ29N7b7q1TTN+yrJLyS5uC3/HPBPGZxr+RLw663bqftpZf/9OvDnNThz3au1byvO1J/tB4MrMP6SwZzdb2/1eDYw7jcwuPLgG8DhlbEzmEt8BHi+PV9SP7ni4ROtzqeB+a2uYaiWzzA4pP+/DP7v5X2j1AG8l8HJtSXgPRNY039pY36q/QPcPtT/t1tNzwG3TOLvJ/CPGUw3PAU82R63TvO+WqOmqd1XwD8A/mcb+zPAv2ntb2DwH/kl4L8BF7T2n22vl9r7bzhdrb2H37iWJHXN4nSTJGlMDAlJUpchIUnqMiQkSV2GhCSpy5CQJHUZEpKkLkNCktT1/wGR4k918iuRogAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pd.Series(slidebase_enh.get_length_distribution()).hist(bins=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f05ab4acda0>"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZMAAAD8CAYAAACyyUlaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAFGtJREFUeJzt3X+MXXWZx/H3s60QgiJFlknTNltc+4coWYQJdMPGzEq2lPpHMYGkhNiCJDUEEk26iVX/gIgksAmSsKvs1tDQGtfKooYmlK0NMjEm/GhRpNQudsSuDG1osBWpRt3is3/c78CZ4c6dmfutvb3T9yu5uec+53u+596HO/PhnHvuNDITSZJq/FWvn4Akqf8ZJpKkaoaJJKmaYSJJqmaYSJKqGSaSpGqGiSSpmmEiSapmmEiSqs3t9RM4Uc4999xcvHhxV9v+7ne/48wzzzy+T2gWsT+Tszed2Z/OTob+PPvss69l5l9PNe6UCZPFixeza9eurrYdHh5maGjo+D6hWcT+TM7edGZ/OjsZ+hMR/zudcZ7mkiRVM0wkSdUME0lSNcNEklTNMJEkVTNMJEnVDBNJUjXDRJJUzTCRJFU7Zb4Bf7wsXv/ouMf77/p4j56JJJ08PDKRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVmzJMImJRRDwREXsjYk9EfKbUb4+IVyLiuXJb0djm8xExEhEvRsSVjfryUhuJiPWN+vkR8XRE7IuIb0fEaaV+enk8UtYvnmofkqQTbzpHJseAdZn5QWApcEtEXFDW3ZuZF5XbNoCybhXwIWA58LWImBMRc4CvAlcBFwDXNea5u8y1BDgC3FTqNwFHMvMDwL1l3KT76LoLkqQqU4ZJZh7MzB+X5TeAvcCCDpusBLZk5h8z85fACHBpuY1k5kuZ+SdgC7AyIgL4GPBw2X4TcHVjrk1l+WHgijJ+sn1IknpgRl9aLKeZPgI8DVwO3BoRq4FdtI5ejtAKmqcam43ydvi8PKF+GfA+4DeZeazN+AVj22TmsYh4vYzvtI/m810LrAUYGBhgeHh4Ji/3LUePHn1r23UXHhu3rts5Z5NmfzSevenM/nTWT/2ZdphExLuB7wCfzczfRsT9wB1Alvt7gE8B0WbzpP1RUHYYT4d1nbZ5u5C5AdgAMDg4mN3+W8rNf4f5honfgL++uzlnk5Ph36k+WdmbzuxPZ/3Un2ldzRUR76IVJN/MzO8CZOarmflmZv4Z+Dpvn2YaBRY1Nl8IHOhQfw04OyLmTqiPm6usfy9wuMNckqQemM7VXAE8AOzNzK806vMbwz4BvFCWtwKrypVY5wNLgGeAncCScuXWabQ+QN+amQk8AVxTtl8DPNKYa01Zvgb4QRk/2T4kST0wndNclwOfBHZHxHOl9gVaV2NdROv00n7g0wCZuSciHgJ+RutKsFsy802AiLgV2A7MATZm5p4y3+eALRHxZeAntMKLcv+NiBihdUSyaqp9SJJOvCnDJDN/RPvPKLZ12OZO4M429W3ttsvMl2hzNVZm/gG4dib7kCSdeH4DXpJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVM0wkSdUME0lSNcNEklTNMJEkVTNMJEnVDBNJUjXDRJJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVM0wkSdUME0lSNcNEklTNMJEkVTNMJEnVDBNJUjXDRJJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVmzJMImJRRDwREXsjYk9EfKbUz4mIHRGxr9zPK/WIiPsiYiQino+IixtzrSnj90XEmkb9kojYXba5LyKi231Ikk686RyZHAPWZeYHgaXALRFxAbAeeDwzlwCPl8cAVwFLym0tcD+0ggG4DbgMuBS4bSwcypi1je2Wl/qM9iFJ6o0pwyQzD2bmj8vyG8BeYAGwEthUhm0Cri7LK4HN2fIUcHZEzAeuBHZk5uHMPALsAJaXdWdl5pOZmcDmCXPNZB+SpB6Y0WcmEbEY+AjwNDCQmQehFTjAeWXYAuDlxmajpdapPtqmThf7kCT1wNzpDoyIdwPfAT6bmb8tH2u0Hdqmll3UOz6d6WwTEWtpnQZjYGCA4eHhKaZt7+jRo29tu+7CY+PWdTvnbNLsj8azN53Zn876qT/TCpOIeBetIPlmZn63lF+NiPmZebCcYjpU6qPAosbmC4EDpT40oT5c6gvbjO9mH+Nk5gZgA8Dg4GAODQ1NHDItw8PDjG17w/pHx63bf313c84mzf5oPHvTmf3prJ/6M52ruQJ4ANibmV9prNoKjF2RtQZ4pFFfXa64Wgq8Xk5RbQeWRcS88sH7MmB7WfdGRCwt+1o9Ya6Z7EOS1APTOTK5HPgksDsiniu1LwB3AQ9FxE3Ar4Bry7ptwApgBPg9cCNAZh6OiDuAnWXclzLzcFm+GXgQOAN4rNyY6T4kSb0xZZhk5o9o/xkFwBVtxidwyyRzbQQ2tqnvAj7cpv7rme5DknTi+Q14SVI1w0SSVM0wkSRVM0wkSdUME0lSNcNEklTNMJEkVTNMJEnVDBNJUjXDRJJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVM0wkSdUME0lSNcNEklTNMJEkVTNMJEnVDBNJUjXDRJJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVM0wkSdUME0lSNcNEklTNMJEkVZsyTCJiY0QciogXGrXbI+KViHiu3FY01n0+IkYi4sWIuLJRX15qIxGxvlE/PyKejoh9EfHtiDit1E8vj0fK+sVT7UOS1BvTOTJ5EFjepn5vZl5UbtsAIuICYBXwobLN1yJiTkTMAb4KXAVcAFxXxgLcXeZaAhwBbir1m4AjmfkB4N4ybtJ9zOxlS5KOpynDJDN/CBye5nwrgS2Z+cfM/CUwAlxabiOZ+VJm/gnYAqyMiAA+Bjxctt8EXN2Ya1NZfhi4ooyfbB+SpB6ZW7HtrRGxGtgFrMvMI8AC4KnGmNFSA3h5Qv0y4H3AbzLzWJvxC8a2ycxjEfF6Gd9pH+NExFpgLcDAwADDw8Mzf5XA0aNH39p23YXHxq3rds7ZpNkfjWdvOrM/nfVTf7oNk/uBO4As9/cAnwKizdik/RFQdhhPh3WdthlfzNwAbAAYHBzMoaGhdsOmNDw8zNi2N6x/dNy6/dd3N+ds0uyPxrM3ndmfzvqpP11dzZWZr2bmm5n5Z+DrvH2aaRRY1Bi6EDjQof4acHZEzJ1QHzdXWf9eWqfbJptLktQjXYVJRMxvPPwEMHal11ZgVbkS63xgCfAMsBNYUq7cOo3WB+hbMzOBJ4BryvZrgEcac60py9cAPyjjJ9uHJKlHpjzNFRHfAoaAcyNiFLgNGIqIi2idXtoPfBogM/dExEPAz4BjwC2Z+WaZ51ZgOzAH2JiZe8ouPgdsiYgvAz8BHij1B4BvRMQIrSOSVVPtQ5LUG1OGSWZe16b8QJva2Pg7gTvb1LcB29rUX6LN1ViZ+Qfg2pnsQ5LUG34DXpJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVM0wkSdUME0lSNcNEklTNMJEkVTNMJEnVDBNJUjXDRJJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVM0wkSdUME0lSNcNEklTNMJEkVTNMJEnVDBNJUjXDRJJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVmzJMImJjRByKiBcatXMiYkdE7Cv380o9IuK+iBiJiOcj4uLGNmvK+H0RsaZRvyQidpdt7ouI6HYfkqTemM6RyYPA8gm19cDjmbkEeLw8BrgKWFJua4H7oRUMwG3AZcClwG1j4VDGrG1st7ybfUiSemfKMMnMHwKHJ5RXApvK8ibg6kZ9c7Y8BZwdEfOBK4EdmXk4M48AO4DlZd1ZmflkZiawecJcM9mHJKlHuv3MZCAzDwKU+/NKfQHwcmPcaKl1qo+2qXezD0lSj8w9zvNFm1p2Ue9mH+8cGLGW1qkwBgYGGB4enmLq9o4ePfrWtusuPDZuXbdzzibN/mg8e9OZ/emsn/rTbZi8GhHzM/NgOcV0qNRHgUWNcQuBA6U+NKE+XOoL24zvZh/vkJkbgA0Ag4ODOTQ01G7YlIaHhxnb9ob1j45bt//67uacTZr90Xj2pjP701k/9afb01xbgbErstYAjzTqq8sVV0uB18spqu3AsoiYVz54XwZsL+veiIil5Squ1RPmmsk+JEk9MuWRSUR8i9ZRxbkRMUrrqqy7gIci4ibgV8C1Zfg2YAUwAvweuBEgMw9HxB3AzjLuS5k59qH+zbSuGDsDeKzcmOk+JEm9M2WYZOZ1k6y6os3YBG6ZZJ6NwMY29V3Ah9vUfz3TfUiSesNvwEuSqh3vq7lmpd2vvP6OD94lSW/zyESSVM0wkSRVM0wkSdUME0lSNcNEklTNMJEkVTNMJEnVDBNJUjXDRJJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVM0wkSdUME0lSNcNEklTNMJEkVTNMJEnVDBNJUrW5vX4C/W7x+kffWt5/18d7+EwkqXc8MpEkVTNMJEnVDBNJUjXDRJJUzTCRJFUzTCRJ1QwTSVI1w0SSVK0qTCJif0TsjojnImJXqZ0TETsiYl+5n1fqERH3RcRIRDwfERc35llTxu+LiDWN+iVl/pGybXTahySpN47Hkck/ZuZFmTlYHq8HHs/MJcDj5THAVcCSclsL3A+tYABuAy4DLgVua4TD/WXs2HbLp9iHJKkH/hKnuVYCm8ryJuDqRn1ztjwFnB0R84ErgR2ZeTgzjwA7gOVl3VmZ+WRmJrB5wlzt9iFJ6oHav82VwPcjIoH/yMwNwEBmHgTIzIMRcV4ZuwB4ubHtaKl1qo+2qdNhH+NExFpaRzYMDAwwPDzc1YscOAPWXXhsynHdzt/vjh49esq+9qnYm87sT2f91J/aMLk8Mw+UX+Y7IuJ/OoyNNrXsoj5tJdw2AAwODubQ0NBMNn/Lv37zEe7ZPXWr9l/f3fz9bnh4mG57O9vZm87sT2f91J+q01yZeaDcHwK+R+szj1fLKSrK/aEyfBRY1Nh8IXBgivrCNnU67EOS1ANdh0lEnBkR7xlbBpYBLwBbgbErstYAj5TlrcDqclXXUuD1cqpqO7AsIuaVD96XAdvLujciYmm5imv1hLna7UOS1AM1p7kGgO+Vq3XnAv+Zmf8dETuBhyLiJuBXwLVl/DZgBTAC/B64ESAzD0fEHcDOMu5LmXm4LN8MPAicATxWbgB3TbIPSVIPdB0mmfkS8Hdt6r8GrmhTT+CWSebaCGxsU98FfHi6+5Ak9YbfgJckVTNMJEnVDBNJUjXDRJJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVM0wkSdUME0lSNcNEklTNMJEkVTNMJEnVDBNJUjXDRJJUzTCRJFUzTCRJ1QwTSVI1w0SSVM0wkSRVM0wkSdUME0lSNcNEklRtbq+fwGyyeP2jby3vv+vjPXwmknRieWQiSapmmEiSqhkmkqRqhokkqZphIkmq1tdhEhHLI+LFiBiJiPW9fj6SdKrq20uDI2IO8FXgn4BRYGdEbM3Mn/X2mbV4mbCkU0k/H5lcCoxk5kuZ+SdgC7Cyx89Jkk5JfXtkAiwAXm48HgUu69Fz6ah5lNLkEYuk2aKfwyTa1HLcgIi1wNry8GhEvNjlvs4FXuty20nF3cd7xp75i/RnlrA3ndmfzk6G/vzNdAb1c5iMAosajxcCB5oDMnMDsKF2RxGxKzMHa+eZrezP5OxNZ/ans37qTz9/ZrITWBIR50fEacAqYGuPn5MknZL69sgkM49FxK3AdmAOsDEz9/T4aUnSKalvwwQgM7cB207ArqpPlc1y9mdy9qYz+9NZ3/QnMnPqUZIkddDPn5lIkk4ShkkHp/Kfa4mI/RGxOyKei4hdpXZOROyIiH3lfl6pR0TcV/r0fERc3JhnTRm/LyLW9Or11IqIjRFxKCJeaNSOWz8i4pLS75GybbtL309ak/Tn9oh4pbyHnouIFY11ny+v9cWIuLJRb/szVy60ebr07dvlopu+EBGLIuKJiNgbEXsi4jOlPrveP5nprc2N1of6vwDeD5wG/BS4oNfP6wS+/v3AuRNq/wKsL8vrgbvL8grgMVrf/VkKPF3q5wAvlft5ZXler19bl/34KHAx8MJfoh/AM8Dfl20eA67q9Ws+Dv25HfjnNmMvKD9PpwPnl5+zOZ1+5oCHgFVl+d+Bm3v9mmfQm/nAxWX5PcDPSw9m1fvHI5PJ+eda3mklsKksbwKubtQ3Z8tTwNkRMR+4EtiRmYcz8wiwA1h+op/08ZCZPwQOTygfl36UdWdl5pPZ+s2wuTFXX5ikP5NZCWzJzD9m5i+BEVo/b21/5sr/ZX8MeLhs3+z1SS8zD2bmj8vyG8BeWn/BY1a9fwyTybX7cy0LevRceiGB70fEs+UvCQAMZOZBaP2AAOeV+mS9mu09PF79WFCWJ9Zng1vLqZqNY6dxmHl/3gf8JjOPTaj3nYhYDHwEeJpZ9v4xTCY35Z9rmeUuz8yLgauAWyLiox3GTtarU7WHM+3HbO3T/cDfAhcBB4F7Sv2U7E9EvBv4DvDZzPxtp6Ftaid9fwyTyU3551pms8w8UO4PAd+jdQri1XJITbk/VIZP1qvZ3sPj1Y/Rsjyx3tcy89XMfDMz/wx8ndZ7CGben9doneqZO6HeNyLiXbSC5JuZ+d1SnlXvH8Nkcqfsn2uJiDMj4j1jy8Ay4AVar3/sCpI1wCNleSuwulyFshR4vRy2bweWRcS8copjWanNFselH2XdGxGxtHw+sLoxV98a+0VZfILWewha/VkVEadHxPnAElofILf9mSufAzwBXFO2b/b6pFf+mz4A7M3MrzRWza73T6+vdDiZb7Suqvg5rStMvtjr53MCX/f7aV1J81Ngz9hrp3Xu+nFgX7k/p9SD1j9U9gtgNzDYmOtTtD5gHQFu7PVrq+jJt2idqvk/Wv8neNPx7AcwSOuX7S+Af6N8obhfbpP05xvl9T9P6xfk/Mb4L5bX+iKNK48m+5kr78lnSt/+Czi91695Br35B1qnnZ4Hniu3FbPt/eM34CVJ1TzNJUmqZphIkqoZJpKkaoaJJKmaYSJJqmaYSJKqGSaSpGqGiSSp2v8DUKEb5RjbOrMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pd.Series(transcription_factors_bed.get_length_distribution()).hist(bins=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f486ce15a90>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD8CAYAAACMwORRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAD9NJREFUeJzt3X+s3Xddx/Hni5VBskIxVi6krXTRLrFhxo1rwRDDLaB0xLT/DNJljg2BRsKm/BAtYqaZ/iGgEonzRxMWxSDXgUgaVjP8savG2LEN2EY3q9cxXFfI5IfVywKj8vaPc+jO7m7v/d7Tc+6Zn/t8JDf5/vjc73mfd77n1W8/Pd9vU1VIktrytEkXIEkaPcNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1KANk3rhzZs31/bt2yf18kP5xje+wQUXXDDpMp7y7FM39qk7e/W4u+666ytV9X0rjZtYuG/fvp0777xzUi8/lLm5OWZmZiZdxlOeferGPnVnrx6X5ItdxjktI0kNMtwlqUGGuyQ1yHCXpAYZ7pLUoBXDPclNSR5J8vmz7E+SDySZT3JPkktHX6YkaTW6XLn/MbBnmf2XATv6PweAPzj3siRJ52LFcK+qfwC+tsyQfcCHquco8Jwkzx9VgZKk1RvFnPsW4KGB9RP9bZKkCRnFHapZYtuS/+t2kgP0pm6Ymppibm5uqBe89+FTZ5Yv3rJpqGMMY2FhYeialzOp9zMu4+pTa+xTdy30aq0/56MI9xPAtoH1rcDJpQZW1SHgEMD09HQNezvxNQdvObP84JXDHWMY47oFelLvZ1y8Vbwb+9RdC71a68/5KKZlDgOv639r5iXAqar60giOK0ka0opX7kk+AswAm5OcAH4VeDpAVf0hcAR4NTAPPAq8flzFSpK6WTHcq+qKFfYX8JaRVSRJOmfeoSpJDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqUKdwT7InyfEk80kOLrH/+5PcluSzSe5J8urRlypJ6mrFcE9yHnAjcBmwE7giyc5Fw34FuLmqLgH2A78/6kIlSd11uXLfBcxX1QNV9RgwC+xbNKaAZ/eXNwEnR1eiJGm1NnQYswV4aGD9BPDiRWN+DfhUkuuAC4BXjqQ6SdJQUlXLD0heA7yqqt7YX78K2FVV1w2MeXv/WL+d5MeADwIvrKrvLDrWAeAAwNTU1ItmZ2eHKvreh0+dWb54y6ahjjGMhYUFNm7cOPLjTur9jMu4+tQa+9RdC70a1ed89+7dd1XV9Erjuly5nwC2Daxv5cnTLm8A9gBU1T8neSawGXhkcFBVHQIOAUxPT9fMzEyHl3+yaw7ecmb5wSuHO8Yw5ubmGLbm5Uzq/YzLuPrUGvvUXQu9WuvPeZc59zuAHUkuTHI+vX8wPbxozH8ArwBI8kPAM4H/HGWhkqTuVgz3qjoNXAvcCtxP71sxx5LckGRvf9g7gDcluRv4CHBNrTTfI0kamy7TMlTVEeDIom3XDyzfB7x0tKVJkoblHaqS1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktSgTuGeZE+S40nmkxw8y5jXJrkvybEkfzbaMiVJq7FhpQFJzgNuBH4COAHckeRwVd03MGYH8C7gpVX19STPHVfBkqSVdbly3wXMV9UDVfUYMAvsWzTmTcCNVfV1gKp6ZLRlSpJWo0u4bwEeGlg/0d826CLgoiT/lORokj2jKlCStHqpquUHJK8BXlVVb+yvXwXsqqrrBsZ8Evg28FpgK/CPwAur6r8WHesAcABgamrqRbOzs0MVfe/Dp84sX7xl01DHGMbCwgIbN24c+XEn9X7GZVx9ao196q6FXo3qc7579+67qmp6pXErzrnTu1LfNrC+FTi5xJijVfVt4AtJjgM7gDsGB1XVIeAQwPT0dM3MzHR4+Se75uAtZ5YfvHK4Ywxjbm6OYWtezqTez7iMq0+tsU/dtdCrtf6cd5mWuQPYkeTCJOcD+4HDi8Z8AtgNkGQzvWmaB0ZZqCSpuxXDvapOA9cCtwL3AzdX1bEkNyTZ2x92K/DVJPcBtwHvrKqvjqtoSdLyukzLUFVHgCOLtl0/sFzA2/s/kqQJ8w5VSWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoM6hXuSPUmOJ5lPcnCZcZcnqSTToytRkrRaK4Z7kvOAG4HLgJ3AFUl2LjHuWcDPAbePukhJ0up0uXLfBcxX1QNV9RgwC+xbYtyvA+8FvjnC+iRJQ+gS7luAhwbWT/S3nZHkEmBbVX1yhLVJkoa0ocOYLLGtzuxMnga8H7hmxQMlB4ADAFNTU8zNzXUqcrF3XHz6zPKwxxjGwsLCWF5vUu9nXMbVp9bYp+5a6NVaf867hPsJYNvA+lbg5MD6s4AXAnNJAJ4HHE6yt6ruHDxQVR0CDgFMT0/XzMzMUEVfc/CWM8sPXjncMYYxNzfHsDUvZ1LvZ1zG1afW2KfuWujVWn/Ou0zL3AHsSHJhkvOB/cDh7+6sqlNVtbmqtlfVduAo8KRglyStnRXDvapOA9cCtwL3AzdX1bEkNyTZO+4CJUmr12Vahqo6AhxZtO36s4ydOfeyJEnnwjtUJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQZ3CPcmeJMeTzCc5uMT+tye5L8k9Sf42yQtGX6okqasVwz3JecCNwGXATuCKJDsXDfssMF1VPwx8DHjvqAuVJHXX5cp9FzBfVQ9U1WPALLBvcEBV3VZVj/ZXjwJbR1umJGk1UlXLD0guB/ZU1Rv761cBL66qa88y/veAL1fVbyyx7wBwAGBqaupFs7OzQxV978OnzixfvGXTUMcYxsLCAhs3bhz5cSf1fsZlXH1qjX3qroVejepzvnv37ruqanqlcRs6HCtLbFvyT4QkPw1MAy9ban9VHQIOAUxPT9fMzEyHl3+yaw7ecmb5wSuHO8Yw5ubmGLbm5Uzq/YzLuPrUGvvUXQu9WuvPeZdwPwFsG1jfCpxcPCjJK4F3Ay+rqm+NpjxJ0jC6zLnfAexIcmGS84H9wOHBAUkuAf4I2FtVj4y+TEnSaqwY7lV1GrgWuBW4H7i5qo4luSHJ3v6w9wEbgY8m+VySw2c5nCRpDXSZlqGqjgBHFm27fmD5lSOuS5J0DrxDVZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktSgTuGeZE+S40nmkxxcYv8zkvx5f//tSbaPulBJUncrhnuS84AbgcuAncAVSXYuGvYG4OtV9YPA+4H3jLpQSVJ3Xa7cdwHzVfVAVT0GzAL7Fo3ZB/xJf/ljwCuSZHRlSpJWo0u4bwEeGlg/0d+25JiqOg2cAr53FAVKklZvQ4cxS12B1xBjSHIAONBfXUhyvMPrLytrOwG0GfjKOF9gjd/PuIy9T42wT9011atz/Jy/oMugLuF+Atg2sL4VOHmWMSeSbAA2AV9bfKCqOgQc6lLYU1GSO6tqetJ1PNXZp27sU3f2avW6TMvcAexIcmGS84H9wOFFYw4DV/eXLwf+rqqedOUuSVobK165V9XpJNcCtwLnATdV1bEkNwB3VtVh4IPAnyaZp3fFvn+cRUuSltdlWoaqOgIcWbTt+oHlbwKvGW1pT0n/b6eU1ph96sY+dWevVinOnkhSe3z8gCQ1yHDvS7ItyW1J7k9yLMnP97e/L8m/JLknyV8mec7A77yr/8iF40leNbnq19bZejWw/xeSVJLN/fUk+UC/V/ckuXQyla+t5fqU5Lr+eXMsyXsHtq+7c2qZz96PJDma5HNJ7kyyq799XZ5Pq1ZV/vSmpp4PXNpffhbwr/Qet/CTwIb+9vcA7+kv7wTuBp4BXAj8O3DepN/HJHvVX99G7x/fvwhs7m97NfBX9O6HeAlw+6Tfw4TPqd3A3wDP6O977no+p5bp06eAywbOobn1fD6t9scr976q+lJVfaa//D/A/cCWqvpU9e66BThK73v+0HvkwmxVfauqvgDM03tUQ/PO1qv+7vcDv8gTb2LbB3yoeo4Cz0ny/LWseRKW6dObgd+sqm/19z3S/5V1eU4t06cCnt0ftonH769Zl+fTahnuS+g/1fIS4PZFu36G3hUDdHssQ/MGe5VkL/BwVd29aNi679Wic+oi4Mf7T1D9+yQ/2h9mn57Yp7cC70vyEPBbwLv6w9Z9n7ow3BdJshH4C+CtVfXfA9vfDZwGPvzdTUv8+rr66tFgr+j15t3A9UsNXWLbuunVEufUBuB76E0pvBO4uf+gPfv0xD69GXhbVW0D3kbvfhpY533qynAfkOTp9E6uD1fVxwe2Xw38FHBl9Sf96PZYhmYt0asfoDdPfHeSB+n14zNJnsc67tVZzqkTwMf70wqfBr5D79kp9umJfboa+O7yR3l8imrd9mk1DPe+/pXTB4H7q+p3BrbvAX4J2FtVjw78ymFgf/8/KrkQ2AF8ei1rnpSlelVV91bVc6tqe1Vtp/cBvLSqvkyvV6/rf8vhJcCpqvrSpOpfK2c7p4BPAC/vj7kIOJ/eQ7HW5Tm1TJ9OAi/rL78c+Lf+8ro8n1ar0x2q68RLgauAe5N8rr/tl4EP0Pv2wl/3H1F/tKp+tnqPYLgZuI/elMRbqup/J1D3JCzZq+rdybyUI/S+4TAPPAq8fvwlPiWc7Zy6CbgpyeeBx4Cr+38jXK/n1Nn69Cbgd9N7GOE3efyJsuv1fFoV71CVpAY5LSNJDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lq0P8B1/CLS/9AMx8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "pd.Series(div_trans_beds_cov_filtered[0].get_length_distribution()).hist(bins=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Application to MS dataset\n",
    "## Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "ms_meta = pd.read_csv('../../ms_enh/metadata/metadata.csv')\n",
    "\n",
    "# Get Healthy bams\n",
    "healthy_bams = ['../../ms_enh/results/snakemake/star/bams/' + x + '.bam' for x in \n",
    "                ms_meta.loc[ms_meta.Main.isin(['ia', 'Healthy_control'])].loc[:,'Petient ID']]\n",
    "\n",
    "# Get ms bams\n",
    "rrms_bams = ['../../ms_enh/results/snakemake/star/bams/' + x + '.bam' for x in \n",
    "               ms_meta.loc[ms_meta.Diagnose == 'RRMS'].loc[:,'Petient ID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../../ms_enh/results/snakemake/star/bams/Mono_85.bam',\n",
       " '../../ms_enh/results/snakemake/star/bams/Mono_125.bam',\n",
       " '../../ms_enh/results/snakemake/star/bams/Mono_129.bam',\n",
       " '../../ms_enh/results/snakemake/star/bams/Mono_136.bam',\n",
       " '../../ms_enh/results/snakemake/star/bams/Mono_138.bam']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rrms_bams"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Detect divergent transcription in MS dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "healthy_div_trans_beds = [get_divergent_transcription_beds(bam) for bam in healthy_bams]\n",
    "healthy_div_trans_merged = [bed.merge(outfolder='../results/from_bam') for bed in healthy_div_trans_beds]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-8d5f69feedc2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mms_div_trans_beds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mget_divergent_transcription_beds\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbam\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mbam\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mms_bams\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mms_div_trans_merged\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mbed\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutfolder\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'../results/from_bam'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mbed\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mms_div_trans_beds\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-17-8d5f69feedc2>\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mms_div_trans_beds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mget_divergent_transcription_beds\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbam\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mbam\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mms_bams\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mms_div_trans_merged\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mbed\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutfolder\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'../results/from_bam'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mbed\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mms_div_trans_beds\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-6-6e35029f6cb5>\u001b[0m in \u001b[0;36mget_divergent_transcription_beds\u001b[0;34m(bam, out_folder)\u001b[0m\n\u001b[1;32m     12\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbed_out\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'w'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m             \u001b[0;32mfor\u001b[0m \u001b[0minterval\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0midentify_divergent_transcription\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moread_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m                 \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0minterval\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Programming/pyNextGen/div_trans_lib.py\u001b[0m in \u001b[0;36midentify_divergent_transcription\u001b[0;34m(oread_gen, distance)\u001b[0m\n\u001b[1;32m     70\u001b[0m     \u001b[0mupstream_oread\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moread_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 72\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mdownstream_oread\u001b[0m \u001b[0;32min\u001b[0m \u001b[0moread_gen\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     73\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mupstream_oread\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrand\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'-'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Programming/pyNextGen/div_trans_lib.py\u001b[0m in \u001b[0;36mget_oriented_reads\u001b[0;34m(bam)\u001b[0m\n\u001b[1;32m     57\u001b[0m                 \u001b[0mlogging\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"{} reads treated\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 59\u001b[0;31m             \u001b[0;32myield\u001b[0m \u001b[0mOread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     60\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Programming/pyNextGen/div_trans_lib.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, ali_read)\u001b[0m\n\u001b[1;32m     24\u001b[0m         \u001b[0mblocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mali_read\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchro\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mali_read\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreference_name\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "ms_div_trans_beds = [get_divergent_transcription_beds(bam) for bam in ms_bams]\n",
    "ms_div_trans_merged = [bed.merge(outfolder='../results/from_bam') for bed in ms_div_trans_beds]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select intervals overlapping with CAGE dataset divergent transcription events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'cage_minus_bed' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-254ac97a35cd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m div_trans_beds_merged_filter = [bed.intersect(cage_minus_bed, supp_args='-wa')                                .intersect(cage_plus_bed, supp_args='-wa')\n\u001b[0;32m----> 2\u001b[0;31m                                 for bed in div_trans_beds_merged]\n\u001b[0m",
      "\u001b[0;32m<ipython-input-16-254ac97a35cd>\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m      1\u001b[0m div_trans_beds_merged_filter = [bed.intersect(cage_minus_bed, supp_args='-wa')                                .intersect(cage_plus_bed, supp_args='-wa')\n\u001b[0;32m----> 2\u001b[0;31m                                 for bed in div_trans_beds_merged]\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'cage_minus_bed' is not defined"
     ]
    }
   ],
   "source": [
    "div_trans_beds_merged_filter = [bed.intersect(cage_minus_bed, supp_args='-wa')\\\n",
    "                                .intersect(cage_plus_bed, supp_args='-wa')\n",
    "                                for bed in div_trans_beds_merged]"
   ]
  }
 ],
 "metadata": {
  "extensions": {
   "jupyter_dashboards": {
    "activeView": "grid_default",
    "version": 1,
    "views": {
     "grid_default": {
      "cellMargin": 10,
      "defaultCellHeight": 20,
      "maxColumns": 12,
      "name": "grid",
      "type": "grid"
     },
     "report_default": {
      "name": "report",
      "type": "report"
     }
    }
   }
  },
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "47px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": false,
   "threshold": 4,
   "toc_cell": false,
   "toc_position": {
    "height": "577px",
    "left": "5px",
    "right": "20px",
    "top": "107.986px",
    "width": "184px"
   },
   "toc_section_display": "block",
   "toc_window_display": true,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
